CHANGELOG.md

```diff
@@ -7,6 +7,7 @@
 - [#1950](https://github.com/openshift/cluster-monitoring-operator/pull/1950) Disable CORS headers on Thanos querier by default and add a flag to enable them back.
 - [#1963](https://github.com/openshift/cluster-monitoring-operator/pull/1963) Add nodeExporter settings for network devices list.
 - [#2049](https://github.com/openshift/cluster-monitoring-operator/pull/2049) Remove Kube*QuotaOvercommit alerts.
+- [#2067](https://github.com/openshift/cluster-monitoring-operator/pull/2067) Add options to specify resource requests and limits for all components.
```

Comment:
Cool, finally :)

but IIUC, these new fields would override the default requests we already set.

Is it really what we want to do? Force users to define both requests and limits even when they only want to set limits?

IIRC, when only limits are defined, Kubernetes will set requests to the same values than limits.

Reply:
1. if we don't set requests for the components, it will use the default requests as defined in each deployment/daemonset
2. https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
```
Note: If you specify a limit for a resource, but do not specify any request, and no admission-time mechanism has applied a default request for that resource, then Kubernetes copies the limit you specified and uses it as the requested value for the resource.
```
3. specify a request/limit, the component should use the specified settings

Reply:
Thanks @juzhao for confirming that "if only limits are set, requests are set equal to limits".

Here are my concerns in two scenarios:

- User only sets the limits via this feature field without knowing that the default requests are not taken into account anymore and that the requests will be equal to the limits they provided (they may have some capacity issues) (unexpected behaviour)
- User took the care to copy the default requests in this feature field, but I a change was done that makes a component need less or more requests, we take care of changing the default requests, but the requests in the feature field are not up to date anymore. (maintenance burden)

Reply:
> Thanks @juzhao for confirming that "if only limits are set, requests are set equal to limits".
> 
> Here are my concerns in two scenarios:
> 
>     * User only sets the limits via this feature field without knowing that the default requests are not taken into account anymore and that the requests will be equal to the limits they provided (they may have some capacity issues) (unexpected behaviour)
I think this is expected behavior, by design of Kubernetes, user should know it

>     * User took the care to copy the default requests in this feature field, but I a change was done that makes a component need less or more requests, we take care of changing the default requests, but the requests in the feature field are not up to date anymore. (maintenance burden)

prometheus-adapter default resources.requests
https://github.com/openshift/cluster-monitoring-operator/blob/release-4.14/assets/prometheus-adapter/deployment.yaml#L73-L76
1. if use does not set resources.requests in cluster-monitoring-config configmap, prometheus-adapter will sue the  default resources.requests(if the default value can not let prometheus-adapter pod start up, then it's a bug, should be fixed in CMO)
2. if user only set resources.limits, resources.requests use the same value as resources.limits,
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    k8sPrometheusAdapter:
      resources:
        limits:
          cpu: 200m
          memory: 500Mi
```
3. set resources.limits and resources.requests, pod should use the  specified value
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    k8sPrometheusAdapter:
      resources:
        requests:
          cpu: 20m
          memory: 200Mi
        limits:
          cpu: 200m
          memory: 500Mi
```
when set resources.limits and resources.requests, user should know clearly of the prometheus-adapter cpu/memory usage, then set resources.limits and resources.requests to a proper value, otherwise, pod would not start up